api/src/main/java/io/minio/MinioClient.java

```diff
@@ -4711,6 +4961,92 @@ public void putObject(String bucketName, String objectName, InputStream stream,
+   *   builder.append("brave ghost pled; Watch Jeopardy!, Alex Trebeks fun TV quiz game.\n");
+   *   builder.append("---\n");
+   * }
+   * ByteArrayInputStream bais = new ByteArrayInputStream(builder.toString().getBytes("UTF-8"));
+   * // create object
+   * Map<String, String> headerMap = new HashMap<>();
+   * headerMap.put("Content-Type", "application/octet-stream");
+   * headerMap.put("X-Amz-Meta-Key", "meta-data");
+
+   * //Generate symmetric 256 bit AES key.
+   * KeyGenerator symKeyGenerator = KeyGenerator.getInstance("AES");
+   * symKeyGenerator.init(256);
+   * SecretKey symKey = symKeyGenerator.generateKey();
+
+   * minioClient.putObject("my-bucketname", "my-objectname", bais, bais.available(), headerMap, symKey);
+   * bais.close();
+   * System.out.println("my-objectname is uploaded successfully"); }</pre>
+   *
+   * @param bucketName  Bucket name.
+   * @param objectName  Object name to create in the bucket.
+   * @param stream      stream to upload.
+   * @param size        Size of all the data that will be uploaded.
+   * @param headerMap   Custom/additional meta data of the object.
+   * @param sse         encryption metadata.
+   * @param contentType Content type of the stream.
+   * @param objectLockRetention Objecct Lock Retention.
+   *
+   * @throws InvalidBucketNameException  upon invalid bucket name is given
+   * @throws NoSuchAlgorithmException
+   *           upon requested algorithm was not found during signature calculation
+   * @throws InsufficientDataException  upon getting EOFException while reading given
+   *           InputStream even before reading given length
+   * @throws IOException                 upon connection error
+   * @throws InvalidKeyException
+   *           upon an invalid access key or secret key
+   * @throws NoResponseException         upon no response from server
+   * @throws XmlPullParserException      upon parsing response xml
+   * @throws ErrorResponseException      upon unsuccessful execution
+   * @throws InternalException           upon internal library error
+   * @throws InvalidArgumentException    upon invalid value is passed to a method.
+   * @throws InvalidResponseException    upon a non-xml response from server
+   */
+  public void putObject(String bucketName, String objectName, InputStream stream, Long size, 
+              Map<String, String> headerMap, ServerSideEncryption sse, String contentType, boolean objectLockRetention)
+    throws InvalidBucketNameException, NoSuchAlgorithmException, IOException,
+           InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+           InternalException, InvalidArgumentException, InsufficientDataException, InvalidResponseException {
+
+    if (!(stream instanceof BufferedInputStream)) {
+      stream = new BufferedInputStream(stream);
+    }
+    if (objectLockRetention) {
+      headerMap.put("retention","");
```

Comment:
Two things you can do at the time of put object.
1. retention (`Mode` and `RetainUntilDate` are required)
2. legal hold (`boolean` to say `ON` or `OFF`)

You could send these changes in put object as separate PR.

Reply:
Will send a separate PR for this